pennylane_lightning/core/src/measurements/MeasurementsBase.hpp

```diff
@@ -111,6 +135,167 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
+                     "supported by shots");
+        } else if (obs.getObsName().find("Hamiltonian") != std::string::npos) {
+            auto coeffs = obs.getCoeffs();
+            for (size_t obs_term_idx = 0; obs_term_idx < coeffs.size();
+                 obs_term_idx++) {
+                auto obs_samples =
+                    samples(obs, num_shots, shot_range, obs_term_idx);
+                size_t num_elements = 0;
+                PrecisionT result_per_term = 0.0;
+                for (int element : obs_samples) {
+                    result_per_term += element;
+                    num_elements++;
+                }
+                result += coeffs[obs_term_idx] * result_per_term / num_elements;
+            }
+        } else {
+            auto obs_samples = samples(obs, num_shots, shot_range);
+            size_t num_elements = 0;
+            for (int element : obs_samples) {
+                result += element;
+                num_elements++;
+            }
+            result = result / num_elements;
+        }
+
+        return result;
+    }
+
+    /**
+     * @brief Return samples of a observable
+     *
+     * @param obs The observable to sample
+     * @param num_shots Number of shots used to generate samples
+     * @param shot_range The range of samples to use. If it's empty, all samples
+     * are used.
+     * @param bin_size  Divides the shot range into bins of size ``bin_size``,
+     * and returns the measurement statistic separately over each bin.
+     * @param counts Whether count("True") or raw samples ("False") should be
+     * retruned
+     *
+     * @return std::vector<size_t> samples in std::vector
+     */
+    auto samples(const Observable<StateVectorT> &obs, const size_t &num_shots,
+                 const std::vector<size_t> &shot_range,
+                 const size_t term_idx = 0,
+                 [[maybe_unused]] size_t bin_size = 0,
+                 [[maybe_unused]] bool counts = false) {
+        std::vector<size_t> obs_wires; // for Hamiltonian this obs_wires should
+                                       // be calculated based on terms
+        const size_t num_qubits = _statevector.getTotalNumQubits();
+
+        bool shots = true;
+        std::vector<size_t> identify_wires;
+
+        StateVectorT sv(_statevector);
+
+        obs.applyInPlace(sv, shots, identify_wires, obs_wires, term_idx);
+
+        Derived measure(sv);
+
+        std::vector<size_t> samples = measure.generate_samples(num_shots);
+        std::vector<size_t> sub_samples;
+        std::vector<PrecisionT> obs_samples(num_shots, 0);
+
+        if (shot_range.empty()) {
+            sub_samples = samples;
+        } else {
+            // Get a slice of samples based on the shot_range vector
+            for (auto &i : shot_range) {
+                for (size_t j = i * num_qubits; j < (i + 1) * num_qubits; j++) {
+                    sub_samples.push_back(samples[j]);
+                }
```

Comment:
We can do some extra work to make it cache-friendly. Let's add a TODO here for the future reference
.